### PR TITLE
Home page view with hardcoded books

### PR DESCRIPTION
**Pull request by petarLits**
Finished for today. There is still work to do.

### CODE REVIEW

File: lib/app_routes.dart [hunk 1]
```diff
@@ -25,6 +27,8 @@ class AppRoutes {
         return _materialRoute(
           SignUp(),
         );
+      case addNewBookRoute:
+        return _materialRoute(AddNewBook());
```
Comment (lecicdjuro) [inline lib/app_routes.dart]: this is not the right place to make instance of new book. You should forward it using params

File: lib/book.dart [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import 'package:book/book_data.dart';
+import 'package:equatable/equatable.dart';
+
+class Book extends Equatable {
+  Book({
+    required this.title,
+    required this.author,
+    required this.imageUrl,
+    this.bookData,
+    required this.docId,
```
Comment (lecicdjuro) [inline lib/book.dart]: required before optional

File: lib/data/firebase_db_manager.dart [hunk 3]
```diff
@@ -2,17 +2,19 @@ import 'package:book/core/constants.dart';
 import 'package:cloud_firestore/cloud_firestore.dart';
 import 'package:firebase_storage/firebase_storage.dart';
 
-import '../book.dart';
+import '../book/book.dart';
```
Comment (lecicdjuro) [inline lib/data/firebase_db_manager.dart]: we should not use relative path for imports. Try to use absolute path

File: lib/home/bloc/home_event.dart [hunk 4]
```diff
@@ -1,5 +1,9 @@
+import 'dart:io';
+
 import 'package:equatable/equatable.dart';
 
+import '../../book/book.dart';
```
Comment (lecicdjuro) [inline lib/home/bloc/home_event.dart]: ditto

File: lib/home/bloc/home_state.dart [hunk 5]
```diff
@@ -1,6 +1,9 @@
 import 'dart:math';
 
 import 'package:equatable/equatable.dart';
+import 'package:image_picker/image_picker.dart';
+
+import '../../book/book.dart';
```
Comment (lecicdjuro) [inline lib/home/bloc/home_state.dart]: ditto

File: lib/home/bloc/home_bloc.dart [hunk 6]
```diff
@@ -26,7 +27,11 @@ class HomeBloc extends Bloc<HomeEvent, HomeState> {
     }
   }
 
-  void _onAddNewBook(AddNewBook event, Emitter<HomeState> emit) {
+  Future<void> _onAddNewBook(AddNewBook event, Emitter<HomeState> emit) async {
     emit(AddingNewBook());
```
Comment (lecicdjuro) [inline lib/home/bloc/home_bloc.dart]: AddNewBookState

File: lib/home/bloc/home_bloc.dart [hunk 7]
```diff
@@ -26,7 +27,11 @@ class HomeBloc extends Bloc<HomeEvent, HomeState> {
...
   }
+
+  Future<void> _onSaveNewBook(SaveNewBook event, Emitter<HomeState> emit)async {
+    emit(SavedBook(book: event.book));
```
Comment (lecicdjuro) [inline lib/home/bloc/home_bloc.dart]: SaveBookState

File: lib/home/view/add_new_book.dart [hunk 8]
```diff
@@ -15,126 +27,146 @@ class _AddNewBookState extends State<AddNewBook> {
...
+                  height: 20,
+                ),
+                Row(
+                  mainAxisAlignment: MainAxisAlignment.center,
+                  children: [
+                    IconButton(
+                      onPressed: () async {
+                        final pickedImage =
+                            await picker.pickImage(source: ImageSource.gallery);
+                        if (pickedImage != null) {
```
Comment (lecicdjuro) [inline lib/home/view/add_new_book.dart]: AddBookImageEvent

Comment (lecicdjuro) [reply lib/home/view/add_new_book.dart]: and forward file

**Review by lecicdjuro — CHANGES_REQUESTED** [review]

File: lib/data/firebase_db_manager.dart [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+      _instance = FirebaseDbManager._internal();
+    }
+    return _instance!;
+  }
+
+  static FirebaseDbManager get instance => FirebaseDbManager();
+
+  Future<void> uploadBookImageAndGetUrl(Book book) async{
+
+    final folderRef= storageRef.child(book.title);
```
Comment (lecicdjuro) [inline lib/data/firebase_db_manager.dart]: format

File: lib/data/firebase_db_manager.dart [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+          throw Exception(serverError);
+        });
+    final url = await fileRef.getDownloadURL().timeout(Duration(seconds: 3),
+        onTimeout: () {
+          throw Exception(serverError);
+        });
+    book.imageUrl = url;
+  }
+
+  Future<void> uploadBook(Book book) async{
```
Comment (lecicdjuro) [inline lib/data/firebase_db_manager.dart]: format

File: lib/data/firebase_db_manager.dart [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    final bookRef = db.collection('books').doc();
+    await bookRef.set(book.toJson()).timeout(Duration(seconds: 3),
+        onTimeout: () {
+          throw Exception(serverError);
+        });
+
```
Comment (lecicdjuro) [inline lib/data/firebase_db_manager.dart]: format

File: lib/data/firebase_db_manager.dart [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+  factory FirebaseDbManager() {
+    if (_instance == null) {
+      _instance = FirebaseDbManager._internal();
+    }
+    return _instance!;
+  }
+
+  static FirebaseDbManager get instance => FirebaseDbManager();
+
+  Future<void> uploadBookImageAndGetUrl(Book book) async{
```
Comment (lecicdjuro) [inline lib/data/firebase_db_manager.dart]: format

File: lib/home/bloc/home_bloc.dart [hunk 13]
```diff
@@ -2,26 +2,76 @@ import 'dart:async';
...
       );
+      emit(LoadedState());
     }
   }
+
+  Future<void> _onUploadBook(UploadBook event, Emitter<HomeState> emit) async {
+
+    emit(LoadingState());
+
+    try{
```
Comment (lecicdjuro) [inline lib/home/bloc/home_bloc.dart]: format

File: lib/home/bloc/home_bloc.dart [hunk 14]
```diff
@@ -2,26 +2,76 @@ import 'dart:async';
...
+      FirebaseDbManager.instance.uploadBook(event.book);
+      emit(LoadedState());
+      emit(UploadedBookState(book: event.book));
+    }on Exception catch (e){
```
Comment (lecicdjuro) [inline lib/home/bloc/home_bloc.dart]: format

File: lib/home/bloc/home_bloc.dart [hunk 15]
```diff
@@ -2,26 +2,76 @@ import 'dart:async';
...
+  }
+
+  Future<void> _onAddBookImage(AddBookImage event, Emitter<HomeState> emit) async{
+    emit(AddBookImageState(image: event.image));
+  }
+
+  Future<void> _onUploadBookImage(UploadBookImageAndGetUrl event, Emitter<HomeState> emit) async{
+
+    emit(LoadingState());
+    try{
```
Comment (lecicdjuro) [inline lib/home/bloc/home_bloc.dart]: format

File: lib/home/bloc/home_bloc.dart [hunk 16]
```diff
@@ -2,26 +2,76 @@ import 'dart:async';
...
+     await FirebaseDbManager.instance.uploadBookImageAndGetUrl(event.book);
+     emit(LoadedState());
+      emit(UploadedBookImageAndUrlGotState(book: event.book));
+    }on Exception catch(e){
```
Comment (lecicdjuro) [inline lib/home/bloc/home_bloc.dart]: format

File: lib/home/bloc/home_bloc.dart [hunk 17]
```diff
@@ -2,26 +2,76 @@ import 'dart:async';
...
+      emit(LoadedState());
+    }
+
+  }
+
+  Future<void> _onSaveNewBook(SaveNewBook event, Emitter<HomeState> emit)async {
+    emit(SaveBookState(book: event.book));
+  }
+
+  Future<void> _onAddBookImage(AddBookImage event, Emitter<HomeState> emit) async{
```
Comment (lecicdjuro) [inline lib/home/bloc/home_bloc.dart]: format

File: lib/home/bloc/home_bloc.dart [hunk 18]
```diff
@@ -2,26 +2,76 @@ import 'dart:async';
...
+      emit(LoadedState());
+      emit(UploadedBookState(book: event.book));
+    }on Exception catch (e){
+      emit(ErrorState(error: e));
+      emit(LoadedState());
+    }
+
+  }
+
+  Future<void> _onSaveNewBook(SaveNewBook event, Emitter<HomeState> emit)async {
```
Comment (lecicdjuro) [inline lib/home/bloc/home_bloc.dart]: format

File: lib/home/bloc/home_event.dart [hunk 19]
```diff
@@ -1,11 +1,56 @@
...
+
+  UploadBook({required this.book});
+
+  final Book book;
+
+  @override
+  List<Object?> get props => [book];
+}
+
+class SaveNewBook extends HomeEvent{
```
Comment (lecicdjuro) [inline lib/home/bloc/home_event.dart]: format

File: lib/home/bloc/home_event.dart [hunk 20]
```diff
@@ -1,11 +1,56 @@
...
+  SaveNewBook({required this.book});
+
+  final Book book;
+
+  @override
+  List<Object?> get props => [book];
+}
+
+class AddBookImage extends HomeEvent{
```
Comment (lecicdjuro) [inline lib/home/bloc/home_event.dart]: format

File: lib/home/bloc/home_event.dart [hunk 21]
```diff
@@ -1,11 +1,56 @@
...
+
+  AddBookImage({required this.image});
+
+  final File image;
+
+  @override
+  List<Object?> get props => [image];
+}
+
+class UploadBookImageAndGetUrl extends HomeEvent{
```
Comment (lecicdjuro) [inline lib/home/bloc/home_event.dart]: format

File: lib/home/bloc/home_event.dart [hunk 22]
```diff
@@ -1,11 +1,56 @@
...
+
+  UploadBookImageAndGetUrl({required this.book});
+
+  final Book book;
+
+  @override
+  List<Object?> get props => [book];
+}
+
+class DeleteBookImage extends HomeEvent{
```
Comment (lecicdjuro) [inline lib/home/bloc/home_event.dart]: format

File: lib/home/bloc/home_state.dart [hunk 23]
```diff
@@ -1,29 +1,88 @@
...
+
+class SaveBookState extends HomeState {
+  SaveBookState({required this.book});
+
+  final Book book;
+  @override
+  List<Object?> get props => [book, Random().nextInt(10000)];
+}
+
+class AddBookImageState extends HomeState{
```
Comment (lecicdjuro) [inline lib/home/bloc/home_state.dart]: format

File: lib/home/bloc/home_state.dart [hunk 24]
```diff
@@ -1,29 +1,88 @@
...
+
+  AddBookImageState({required this.image});
+
+  final File image;
+}
+
+class UploadedBookImageAndUrlGotState extends HomeState{
```
Comment (lecicdjuro) [inline lib/home/bloc/home_state.dart]: ditto

File: lib/home/bloc/home_state.dart [hunk 25]
```diff
@@ -1,29 +1,88 @@
...
+
+  UploadedBookImageAndUrlGotState({required this.book});
+
+  final Book book;
+
+  @override
+  List<Object?> get props => [book];
+}
+
+class DeletedBookImage extends HomeState{
```
Comment (lecicdjuro) [inline lib/home/bloc/home_state.dart]: ditto

File: lib/home/bloc/home_state.dart [hunk 26]
```diff
@@ -1,29 +1,88 @@
...
+
+  DeletedBookImage();
+
+  @override
+  List<Object?> get props => [Random().nextInt(10000)];
+}
+
+class LoadingState extends HomeState{
```
Comment (lecicdjuro) [inline lib/home/bloc/home_state.dart]: ditto

File: lib/home/view/add_new_book.dart [hunk 27]
```diff
@@ -0,0 +1,213 @@
...
+  final _formKey = GlobalKey<FormState>();
+  late String titleValue;
+  late String authorValue;
+  ImagePicker picker = ImagePicker();
+  File? image;
+
+  @override
+  Widget build(BuildContext context) {
+    return BlocConsumer<HomeBloc, HomeState>(listener: (context, state) {
+      if (state is SaveBookState) {
```
Comment (lecicdjuro) [inline lib/home/view/add_new_book.dart]: SavedBookState

File: lib/home/view/add_new_book.dart [hunk 28]
```diff
@@ -0,0 +1,213 @@
...
+        image = state.image;
+      } else if (state is UploadedBookImageAndUrlGotState) {
+        context.read<HomeBloc>().add(SaveNewBook(book: state.book));
+      } else if (state is ErrorState) {
+        final snackBar = SnackBar(
+          content: Text(AppLocalizations.of(context)!.serverError),
+          backgroundColor: AppColors.errorSnackBar,
+        );
+        ScaffoldMessenger.of(context).showSnackBar(snackBar);
+      }else if(state is DeletedBookImage){
```
Comment (lecicdjuro) [inline lib/home/view/add_new_book.dart]: fkr at

File: lib/home/view/home_page.dart [hunk 29]
```diff
@@ -16,34 +16,192 @@ class HomePage extends StatefulWidget {
...
+                    ),
+                  ),
+                  Spacer(),
+                  Row(
+                    mainAxisAlignment: MainAxisAlignment.end,
+                    children: [
+                      Padding(
+                        padding: EdgeInsets.only(right: 24),
+                        child: Text(
+                          '${AppLocalizations.of(context)!.author} ${book.author}',
```
Comment (lecicdjuro) [inline lib/home/view/home_page.dart]: use localization with params

File: lib/login/bloc/login_bloc.dart [hunk 30]
```diff
@@ -22,16 +24,21 @@ class LoginBloc extends Bloc<LoginEvent, LoginState> {
     } on Exception catch (e) {
       emit(ErrorState(error: e));
     }
+    finally{
```
Comment (lecicdjuro) [inline lib/login/bloc/login_bloc.dart]: format

File: lib/login/view/login_page.dart [hunk 31]
```diff
@@ -56,6 +60,10 @@ class _LoginPageState extends State<LoginPage> {
           content: Text(AppLocalizations.of(context)!.serverError),
         );
         ScaffoldMessenger.of(context).showSnackBar(snackBar);
+      }else if(state is LoadingState){
```
Comment (lecicdjuro) [inline lib/login/view/login_page.dart]: format

File: lib/login/view/login_page.dart [hunk 32]
```diff
@@ -56,6 +60,10 @@ class _LoginPageState extends State<LoginPage> {
...
+        DialogUtils.showLoadingScreen(context);
+      }else if(state is LoadedState){
```
Comment (lecicdjuro) [inline lib/login/view/login_page.dart]: ditto

**Review by lecicdjuro — CHANGES_REQUESTED** [review]

File: lib/app_user.dart [hunk 33]
```diff
@@ -19,10 +19,11 @@ class AppUser extends Equatable {
...
 
-  factory AppUser.fromJson(Map<String,dynamic> json) => _$AppUserFromJson(json);
+  factory AppUser.fromJson(Map<String, dynamic> json) =>
+      _$AppUserFromJson(json);
 
   Map<String, dynamic> toJson() => _$AppUserToJson(this);
 
   @override
-  List<Object?> get props =>[firstName, lastName, email, password, isAdmin];
+  List<Object?> get props => [firstName, lastName, email, password, isAdmin];
```
Comment (milicaLits) [inline lib/app_user.dart:27]: Add comma after last parameter and format

File: lib/home/bloc/home_bloc.dart [hunk 34]
```diff
@@ -18,7 +19,7 @@ class HomeBloc extends Bloc<HomeEvent, HomeState> {
   }
 
   Future<void> _onSignOut(SignOut event, Emitter<HomeState> emit) async {
-   emit(LoadingState());
+    emit(LoadingState());
 
     try {
       await FirebaseAuthManager.instance.signOut();
```
Comment (milicaLits) [inline lib/home/bloc/home_bloc.dart:25]: No comma when only one parameter. Format.

File: lib/home/bloc/home_state.dart [hunk 35]
```diff
@@ -38,22 +38,24 @@ class UploadedBookState extends HomeState {
   List<Object?> get props => [book];
 }
 
-class SaveBookState extends HomeState {
-  SaveBookState({required this.book});
+class SavedBookState extends HomeState {
+
```
Comment (milicaLits) [inline lib/home/bloc/home_state.dart]: Format

**Review by milicaLits — APPROVED** [review]

**Review by lecicdjuro — APPROVED** [review]

**Pull request merged**